mypy/checker.py

```diff
@@ -280,38 +320,57 @@ def get_generator_yield_type(self, return_type: Type) -> Type:
             # be accessed so any type is acceptable.
             return AnyType()
 
-    def get_generator_receive_type(self, return_type: Type) -> Type:
+    def get_generator_receive_type(self, return_type: Type, is_coroutine: bool) -> Type:
+        """Given a declared generator return type (t), return the type its yield receives (ts)."""
         if isinstance(return_type, AnyType):
             return AnyType()
-        elif not self.is_generator_return_type(return_type):
+        elif not self.is_generator_return_type(return_type, is_coroutine):
             # If the function doesn't have a proper Generator (or superclass) return type, anything
             # is permissible.
             return AnyType()
         elif not isinstance(return_type, Instance):
             # Same as above, but written as a separate branch so the typechecker can understand.
             return AnyType()
         elif return_type.type.fullname() == 'typing.Generator':
-            # Generator is the only type which specifies the type of values it can receive.
-            return return_type.args[1]
+            # Generator is one of the two types which specify the type of values it can receive.
+            if len(return_type.args) == 3:
+                return return_type.args[1]
+            else:
+                return AnyType()
+        elif return_type.type.fullname() == 'typing.Awaitable':
+            # Generator is one of the two types which specify the type of values it can receive.
```

Comment:
Oh, yeah.

Most code shouldn't be sending values into Awaitables -- only the coroutine runner should do that. When using asyncio, the type of values sent into the Awaitable is always asyncio.Future, but it's up to the framework you use. It'd be a distracting detail for users who write coroutines, so the Awaitable type doesn't mention this.